Feature/us17/creation page contact

Samaze123 opened the pull request:
# US 17

Ajout d'une page contact qui permet a des utilisateurs, connectés ou pas, de prendre contact avec les développeurs du site. Les messages des utilisateurs sont appelés tickets et seulement certains utilisateurs seront mis en tant qu'administrateurs ,grâce a un champs "ability" dans la base de données, et pourront voir ces tickets. J'ai créé une page exclusivement réservé aux administrateurs.

## Ajout de la page contact dans le backend

### Ajout des fonctionnalité indispensable coté backend

- app.js : Ajout de la route contact dans le routeur
- tickets.js : Ajout du modèle de ticket de la base de données
- user.js : Ajout de l'attribut ability dans le User ce qui permet "d'être" admin des tickets et d'avoir les droits de :
  - visionner les tickets ouverts.
  - supprimer les tickets.
- auth.js : Ajout d'une réponse lors du login qui sera stockée dans le localstorage du client pour éviter de devoir faire plusieurs requêtes similaires.
- contact.js : Creation de la route :
  - Ajouter, supprimer et recevoir les tickets.
  - Ajout d'un get avec un Id particulier pour avoir plus de controle si on le souhaite.

#### Tests Backend

- contact.test.js : Ajouts de plusieurs tests sur la route /contact.

## Ajout de la page Contact dans le front

### Cette page permet aux utilisateur qui le souhaite de faire parvenir chez les développeur des idées ou autres

- app.tsx : Ajout de la page contact et ticket viewer dans le routeur du front
- ContactUS.tsx : Création de la page contact

## Ajout de la page tickets Viewer

### Cette page sert a visionner tous les tickets ouverts par les utilisateurs

#### Pour l'instant n'a pas de pagination ce qui pourrait créer des problemes dans le futur

- tickets.tsx : Ajout du composant Ticket pour une réutilisation ultérieure plus facile
- TicketsViewer.tsx : Création de la page admin de lecture des tickets.

## Autre

### Ajouts divers utile au reste du site

- Ajout de la réponse du login recue du back dans le local storage
- Dashboard.tsx : Ajout d'un lien vers la page contact dans le dashboard car pas de navbar utilisable
- components/index.ts : Ajout des nouveaux composants dans la liste de composants
- Navbar.tsx : Ajout de lien vers les différente page et mise a jour de la navbar en fonction de l'état de l'utilisateur.
- main.tsx : Ajout d'un userProvider et userContext pour avoir accès a plusieurs données dans le front sans faire de requêtes.
- userContext.tsx : Ajout du context dans un composant réutilisable
- userProvider.tsx : Ajout du rovider dans un composant réutilisable mis sur la racinde du website
- About.tsx : Ajout de la navbar dans la page About
- pages/index.ts : Ajout des pages dans la liste d'import
- DOSSIER /types : Ajout de fichiers contenant les différents types utilisés dans certains modules .TSX.
- Changements mineurs dans différents fichier qui importent leur types depuis le dossier créé.

Julien00859 commented:
Bel effort pour le changelog, vous me présentez ça jeudi je serai content. Pour encore améliorer le style, on comprend en lisant l'ensemble des points qu'il s'agit d'une page contact avec un système (léger) de ticketing, je pense que ce serait intéressant d'avoir une introduction (avant l'énumération des changements) qui vienne expliquer sur 1/2 paragraphe la fonctionnalité en générale, le "pourquoi" c'est utile pour le client.

Je ne vois qu'un test unitaire de 30 lignes pour 600 lignes de code, je trouve ça plutôt léger. C'est bien de vérifier que la route retourne bien une réponse 200, c'est mieux de vérifier que les infos ont été correctement enregistrées en base. Pensez peut-être aussi à rajouter un test pour l'interface, par exemple qui viendrait vérifier qu'on ne peut pas soumettre un formulaire si tous les champs obligatoires ne sont pas remplis.

Samaze123 commented:
> Bel effort pour le changelog, vous me présentez ça jeudi je serai content. Pour encore améliorer le style, on comprend en lisant l'ensemble des points qu'il s'agit d'une page contact avec un système (léger) de ticketing, je pense que ce serait intéressant d'avoir une introduction (avant l'énumération des changements) qui vienne expliquer sur 1/2 paragraphe la fonctionnalité en générale, le "pourquoi" c'est utile pour le client.
> 
> Je ne vois qu'un test unitaire de 30 lignes pour 600 lignes de code, je trouve ça plutôt léger. C'est bien de vérifier que la route retourne bien une réponse 200, c'est mieux de vérifier que les infos ont été correctement enregistrées en base. Pensez peut-être aussi à rajouter un test pour l'interface, par exemple qui viendrait vérifier qu'on ne peut pas soumettre un formulaire si tous les champs obligatoires ne sont pas remplis.

- Ajout d'une explication fait.
- Ajout de tests back et front fait.

Changement de deux trois fonctionnalités expliquées dans le message de commit.

Samaze123 commented:
Suite a un échange avec le groupe, j'ai changé la route en mettant un middleware et j'ai retiré ability du body pour cette requete, c'est le serveur qui va faire la requete pour voir si l'email recu est celle d'un admin ou pas

Samaze123 commented:
Retrait des attribut style et remplacement par du tailwind

JarodEphec commented on g-sonde/src/pages/TicketsViewer.tsx:
```diff
@@ -0,0 +1,106 @@
+				withCredentials: true,
+			}
+		);
+		if (response.status !== 200) {
+			logout();
+			window.location.href = '/login';
+			return;
```
Ce code ne fonctionne pas, quand je ne suis pas autorisé, un toast apparaît avec un page vide sauf la navbar, mais je ne suis pas déconnecté et je ne suis pas amené sur la page de login.

Samaze123 replied:
Code réparé ! Merci du retour !!

JarodEphec commented on g-sonde/src/components/UserProvider.tsx:
```diff
@@ -0,0 +1,57 @@
+import { UserContext } from '../components';
+import { UserProviderProps } from '../types';
+
+const UserProvider: React.FC<UserProviderProps> = ({ children }) => {
+	const [emailUser, setEmailUser] = useState<string | null>(
+		localStorage.getItem('emailUser') || null
+	);
+	const [abilityUser, setAbilityUser] = useState<string | null>(
+		localStorage.getItem('abilityUser') || null
+	);
```
Nous avions parlé de l'utilisation du local storage et du problème de sécurité qu'il peut amener dans ce cas. Tu as bien fait les changements, pas de soucis, mais je vois toujours du local storage dans ton code, peux-tu m'expliquer la raison ?

Samaze123 replied:
J'ai laissé le local storage seulement pour plus de facilité du coté du frontend. Ce qui se trouve dedans sert juste au front qu'est-ce qu'il faut affiché ou pas, ils ne sont plus utilisés dans les requêtes. Est-ce que ca te va ou tu préfère que je retire quand même ?

JarodEphec replied:
Enfaite, si une requête est faite au back end pour faire ce que ce code faisait, je ne vois pas très bien la raison de le garder, ces variables en local storage peuvent-elles être remplacées par des variables ?

Samaze123 replied:
oui oui en vrai je vais retiré le local storage.

Samaze123 replied:
En fait, je stocke ces données dans le localStorage juste pour pouvoir ne pas demandé au back a chaque changement de page. C'est plus performant au pris de sécurité, et encore le seul truc qui pourrait pose problème si quelqu'un arrivait a changer son localStorage, c'est qu'il verrait des boutons qu'il n'est pas senser voir. Mais au moment ou il cliquerait dessus il serait deconnecté et renvoyer au login. 

Du coup si je les remplace par des variable simples, la valeur dans celles-ci serait supprimé z chaque changement de pages et on perdrait cet ajout de perfomance, etc.

JarodEphec reviewed (changes requested):
Salut Samuel, bon travail pour cette pull request !

J'ai quelques remarques à faire.

1. Lorsque j'envoie un ticket et que tout à fonctionner, je n'ai pas de retour au niveau de la page, il faudrait ajouter un retour visuel pour montrer à l'utilisateur que son ticket a bien été envoyé. Je vois dans le code qu'il y a un toast pour une erreur, mais pas pour une réussite, le toast serait un bon retour visuel, mais je trouve qu'une page montrant a l'utilisateur que son ticket a bien été envoyé avec son id de ticket, serait vraiment le summum du retour visuel utilisateur.

2. Quand je suis arrivé sur la page pour voir les tickets, il y avait juste les champs que l'utilisateur avait rentrés (nom, description) et l'id du ticket, mais pas d'information pour contacter l'utilisateur. Lors de la première itération de cette pour request, tu avais fait de sorte que le client mail s'ouvre automatiquement, avec des informations préremplies pour l'utilisateur. Je trouve que ici, ça serait intéressent de remettre cette fonctionnalité, mais cette fois si pour l'admin.
3. Je vois que la seule option disponible pour les tickets est de les supprimer, je pense qu'il serait plus intéressent de les archivés pour avoir un historique des tickets envoyé et comme ça, d'avoir une trace des différents problèmes que les utilisateurs ont lors de l'utilisation de notre solution.
4. Pour l'ouverture de tickets, ça serait bien d'avoir des catégories pour pouvoir trier les tickets et en plus, en mettant des catégories spécifiques, ça rendra ton US encore plus spécifique.
5. Je ne vois pas de test pour la page `/admin/tickets`.
6. Il y a dû refactoring de code dans ta pull request, ça passe ici car c'est du code que tu utilises dans ton US, mais fait attention dans le futur.

bahuwart commented on backend/routes/contact.js:
```diff
@@ -0,0 +1,75 @@
+	}
+});
+
+router.get('/ability', requireUser, async (req, res) => {
+	try {
+		const user = await User.findOne({ email: req.user.email }).select(
+			'ability'
+		);
+		res
+			.status(200)
+			.json({ message: 'User ability sent', ability: user.ability });
+	} catch (e) {
+		console.error(e);
+		res.status(500).json({ message: e });
+	}
+});
```
Salut sam, tu dis dans la pull request que tu as finalement abandonné le champs "ability" mais là je vois que il y a une route ability. J'imagine qu'elle va être enlevée aussi ?
Sinon super boulot : )

JarodEphec replied:
Non, justement, si tu regardes le commit 1f150587e92511a3f2ba9a47a168f82f7676dba7, tu peux voir qu'il a ajouté la route `/ability` justement pour vérifier que l'utilisateur ait bien le droit de voir les tickets.

Samaze123 replied:
On disait que le localStorage n'était pas sécurisé et c'est vrai donc ca il fallait le supprimer mais je garde la partie ability pour savoir si un user est admin ou pas. Juste, la confirmation ne viendra pas du localStorage mais bien du serveur et de ses vérifications

bahuwart replied:
ah ok compris lol

bahuwart reviewed (commented):
J'ai pas grand chose à dire de plus que Jarod.
Beau travail UwU

Samaze123 commented:
J'ai fini normalement d'ajouter ce qu'on a conseillé !

JarodEphec commented on g-sonde/src/pages/CategoryManager.tsx:
```diff
@@ -0,0 +1,247 @@
+			setShowToastError(true);
+			return;
+		}
+		if (!/^[a-zA-Z\s]+$/.test(insertedLabelToFind)) {
+			setToastErrorMessage("Label must only contain letters.");
+			setShowToastError(true);
+			return;
+		}
```
J'ai remarqué que ici, tu n'autorises pas les espaces, ce qui est dérangeant pour écrire une catégorie de problème comme "Problème de conexion avec la sonde".

Samaze123 replied:
Je viens de commit les changements:
- Ajout d'un style qui met a la ligne les mots ou phrases trop longue. (pour le radio bouton)
- Ajout des espaces dans les regex.

JarodEphec reviewed (changes requested):
![Capture d'écran 2023-11-20 214709](https://github.com/Martin-Hayot/Projet-Integration-2023/assets/92374656/57bbb933-1610-4b39-b5aa-cac67f3a75f0)

Ici, on peut voir que ma catégorie avec un nom trop long fait buguer le bouton radio, je sais que tu utilises le composant de ionic, sais-tu faire quelque chose pour ça ?

Sinon, ton US est vraiment impressionnant pour le travail fourni, la beauté de tes pages et les fonctionnalités avancées qu'elles apportent.

JarodEphec reviewed (approved):
Parfait !

Samaze123 committed:
US17

US perso de Samuel Lambert.
Creation de la page contact. Cette page est necessaire pour que l'utilisateur qui le souhaite puisse prendre contact avec l'equipe de la sonde.
Lors du contact avec l'equipe, le message de l'utilisateur est sauve dans un ticket.
Pour que l'equipe puisse lire et manager ces tickets, il a fallu creer une page, uniquement pour l'equipe.
Ensuite, pour plus de classification de ces tickets j'ai cree les categories qui sont manager exclusivement par l'equipe mais qui sont utilisable par l'utilisateur.

BACKEND

Modifications Backend

    Routes

    - index.ts : Ajout de toutes les routes dans ce fichier pour facilite l'import
    user.js : Ajout d'une route pour recevoir l'ability d'un utilisateur.
    - auth.js : Ajout d'une reponse avec des donnees de l'utilisateur qui seront stockee dans le localstorage du front. Ajout de ability lors de la creation d'un compte.

    - category.js : Creation de la route pour ajouter, supprimer et recevoir les categories.
      - Il y a aussi une route qui sert a recevoir seulement une categorie grace a son ID.

    - ticket.js : Creation de la route pour ajouter, supprimer et recevoir les tickets.
      - Utilisation du middleware RequireAdmin et requireUser utilise pour certaines url.
      - Details lors du retour si mauvaises donnees entrees.
      - Il y a aussi une route qui sert a recevoir seulement un ticket grace a son ID.
      - Un utilisateur peut avoir max 5 tickets ouverts en même temps.

    Middleware

    - index.ts : Ajout de tous les middleware dans ce fichier pour facilite l'import.
    - auth.js : Ajout du middleware requireAdmin qui verifie que l'utilisateur est connecte et possede les bons droits admin.

    Modeles

    - index.ts : Ajout de tous les modeles dans ce fichier pour facilite l'import.
    - tickets.js : Ajout du modele Ticket qui correspond a celui de la base de donnees.
    - category.js : Creation du modele Category
    - user.js : Ajout de l'attribut ability dans le User ce qui permet d'etre admin des tickets et d'avoir les droits de :
      - visionner les tickets ouverts.
      - supprimer les tickets.

    Autre

    - app.js : Ajout de la route contact et category dans le routeur backend.

Tests Backend - cypress.config.ts : Ajout d'un parametre pour que les tests prennent des videos. - ticket.test.js : Creation de test pour la route /api/tickets. - Il y a aussi des petits tests en rapport avec les categories.
Pour effectuer les tests backend, il faut utiliser la commande : npm run test

FRONTEND

Modification Frontend

    Composants

    - index.ts : Ajout de tous les composants dans ce fichier pour facilite l'import.
    - Ticket.tsx : Creation du composant Ticket qui permet de l'utiliser plus facilement plusieurs fois.
      - Peut avoir l'attribut readonly pour ne pas avoir de boutons.
    - Navbar.tsx : Ajout d'un bouton pour acceder a la nouvelle page contact.
    - UserProvider.tsx : Ajout du provider pour concentrer toutes la variables "globales" (et inscrites dans le local storage) du site.
    - UserContext.tsx : Ajout du UserContext qui permet d'acceder au UserProvider.
    - Navbar.tsx : Ajout de liens vers les nouvelles pages et changement de navbar en fonction des droits et de l'etat de l'utilisateur.
    - components/Popup.tsx : Ajout d'un popup pour afficher a l'utilisateur qu'il a bien envoye son ticket
    - CategoryCard.tsx : Ajout de la carte Categorie pour plus de facilite lors de son utilisation multiple.

    Pages

    - index.ts : Ajout de toutes les pages dans ce fichier pour facilite l'import.
    - About.tsx : Ajout de la navbar dans la page.
    - CategoryManager : Creation de la page admin pour gerer les categories.
      - La page ne s'affiche seulement lorsque les droits de l'utilisateur ont ete checke auparavant.
    - ContactUs.tsx : Creation de la page contact.
      - Condition avec les inputs et toast error si ces tests ne sont pas valides.
      - Pour acceder a cette page, il faut cliquer sur la bouton dans la navbar ou aller sur LeNomDuSite/contact.
    - Dashboard.tsx : Ajout d'un lien vers la page contact car pas de navbar utilisable.
    - Login.tsx : Ajout de la reponse lors du login dans le local storage pour sauvegarder certains donnees insensibles. (Les donnees sauvegardee dans le local storage ne sont pas utilisees pour les requetes mais seulement afficher plus facilement certaines parties du frontend)
      - Ajout d'un toast pour signaler une erreur lors d'une redirection forcee a l'utilisateur.
    - TicketsViewer.tsx : Creation de la page admin pour manager les tickets (suppression, archivage et prise de contact avec l'utilisateur par mail)
      - La page ne s'affiche seulement si les droits ont ete checke auparavant.
      - Lors de la suppression ou le changement d'etat d'un ticket, la page est mise a jour sans etre rafraichie.
      - Details lors de la suppression d'un ticket.
      - Pas de pagination pour cette page. S'il y a beaucoup de tickets a charger d'un seul coup, peut poser problemes.
      - Bouton qui redirige vers le manager des categories exlusivement reserve aussi aux admins.
    - UserTicketsManager.tsx : Page utilisée par l'utilisateur pour voirs ses tickets ouverts. Les tickets dans cette pages sont en readonly.

    Types

    Ce dosser contient toutes les interfaces utilisees par certains composants et variables
    - index.ts : Ajout de tous les types dans ce fichier pour facilite l'import.

    Autres

    - app.tsx : Ajout de la page Contact, TicketViewer et CategoryManager dans le router du site.
    - main.tsx : Ajout du UserProvider dans l'application pour le mettre accessible a toutes la pages.

Tests Frontend

    Tests End-To-End

    - contact.cy.ts : Contient tous les tests e2e pour la page contact
      - Soumission du formulaire avec un probleme dans le nom de l'utilisateur.
      - Soumission du formulaire sans erreurs.
      - Soumission du formulaire avec des cases vides.

    Pour effectuer les tests unitaire il faut utiliser la commande : npm run test.e2e

    Tests unitaires

    - App.test.tsx : Ajout du test de l'app dans le dossier qui sera utilise lors de la commande.

    Pour effectuer les tests unitaire il faut utiliser la commande : npm run test.unit
